### PR TITLE
feat: rework NodeJS SDK using napi

### PR DESCRIPTION
**Pull request by eddyxu**
Use Napi to write a Node.js SDK that follows Polars for better maintainability, while keeping most of the logic in Rust.

### CODE REVIEW

File: nodejs/vectordb/connection.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+export class Connection {
+  readonly inner: _NativeConnection;
+
+  constructor(inner: _NativeConnection) {
+    this.inner = inner;
+  }
+
+  tableNames = async (): Promise<string[]> => {
+    return this.inner.tableNames();
+  };
```
Comment (westonpace) [inline nodejs/vectordb/connection.ts:34]:
```suggestion
  async tableNames(): Promise<string[]> {
    return this.inner.tableNames();
  };
```
Use a consistent style for function declaration?

File: nodejs/vectordb/connection.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * Connection could be local against filesystem or remote against a server.
+ */
+export class Connection {
+  readonly inner: _NativeConnection;
+
+  constructor(inner: _NativeConnection) {
+    this.inner = inner;
+  }
+
```
Comment (westonpace) [inline nodejs/vectordb/connection.ts:31]:
```suggestion

/**
 * List the names of all tables in the database
 */
```

File: nodejs/vectordb/index.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+ * Connect to a LanceDB instance at the given URI.
+ *
+ * Accpeted formats:
+ *
+ * - `/path/to/database` - local database
+ * - `s3://bucket/path/to/database` or `gs://bucket/path/to/database` - database on cloud storage
+ * - `db://host:port` - remote database (LanceDB cloud)
+ *
+ * @param uri The uri of the database. If the database uri starts with `db://` then it connects to a remote database.
+ *
```
Comment (westonpace) [inline nodejs/vectordb/index.ts:33]: Document the `apiKey` and `hostOverride` parameters?

File: nodejs/src/connection.rs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    ///
+    #[napi]
+    pub async fn create_table(&self, name: String, buf: Buffer) -> napi::Result<Table> {
+        let batches = ipc_file_to_batches(buf.to_vec())
+            .map_err(|e| napi::Error::from_reason(format!("Failed to read IPC file: {}", e)))?;
+        let tbl = self
+            .conn
+            .create_table(&name, Box::new(batches), None)
+            .await
+            .map_err(|e| napi::Error::from_reason(format!("{}", e)))?;
```
Comment (westonpace) [inline nodejs/src/connection.rs:64]: Maybe add a `From` for `napi::Error` so you can use `?`

File: nodejs/vectordb/table.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  async countRows(): Promise<bigint> {
+    return await this.inner.countRows();
+  }
+
+  /** Delete the rows that satisfy the predicate. */
+  async delete(predicate: string): Promise<void> {
+    await this.inner.delete(predicate);
+  }
+
+  search(vector?: number[]): Query {
```
Comment (westonpace) [inline nodejs/vectordb/table.ts:53]: Not documented

**Review by westonpace — APPROVED** [review]
Some nits, but looks easier to understand than what we have today

File: nodejs/src/connection.rs (hunk 4 above)
Comment (eddyxu) [reply nodejs/src/connection.rs:64]: Sg. lemme do it in a follow up.

File: nodejs/vectordb/connection.ts (hunk 1 above)
Comment (eddyxu) [reply nodejs/vectordb/connection.ts:34]: Done

File: nodejs/vectordb/connection.ts (hunk 2 above)
Comment (eddyxu) [reply nodejs/vectordb/connection.ts:31]: done

File: nodejs/vectordb/table.ts (hunk 5 above)
Comment (eddyxu) [reply nodejs/vectordb/table.ts:61]: this interface is not done yet. lemme doc once i am happy with the implementation?

**Pull request merged**